### PR TITLE
Scal xyb

### LINKED ISSUE #557: double precision is needed for high resolution applications
**Describe the bug**
XYB is a variable that keeps the XY coordinates and the depths (all single precision). It does not make a difference in large-scale applications and for very coarse cases. However, if there is a case that requires high resolution in the coastal region, the single-precision is not enough in lat/long coordinates. 

**To Reproduce**
The coastal applications needed double precison. The scalability branch has solved this problem.,

This array should be deleted for the sake of memory saving and towards the integration of the scalability branch into develop.

### PR DESCRIPTION
**Pull request by aronroland**
# Pull Request Summary
A short overview of the PR 
Remove XYB and replaced by double precision XGRD, YGRD and single precision ZB
https://github.com/NOAA-EMC/WW3/projects/2
## Description

XYB is keeping the XY-coordinates and the depths, this array is deleted for the sake of memory saving and towards the integration of the scalability branch into develop. We rather use XGRD, YGRD and ZB instead. XGRD and YGRD are converted to double precision since this is needed by the unstructured part of WW3 for the accurate calculation of the geometric quantities of the unstructured triangular grid. The changes have been implemented in such a way that the structured part of the code and the non-decomposed part of the unstructured grid remains constant. The use of single precision in lat/lon coordinates may also be a bug for the structured, curvilinear, triangular or smc grid if applied to very high resolution.  

Please also include the following information: 
* Add any suggestions for a reviewer 
 _enhancement_
* Are answer changes expected from this PR? Please describe the changes and the reason why in addition to which of the _mod_def change_ only for the unstructured part

### Issue(s) addressed

* Please list any issues associated with this PR, including those the PR will fix/close. For example:  
- fixes #557

### Commit Message

Remove XYB and replaced by double precision XGRD, YGRD and single precision ZB

### Check list  
- [X] Branch is up to date with the authoritative repository (NOAA-EMC) develop branch. 
- [X] Checked the [checklist for a developer submitting to develop](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-a-developer-submitting-to-develop). 
- [x] If a version number update is required, checked the [updating version number](https://github.com/NOAA-EMC/WW3/wiki/Code-Management#checklist-for-updating-version-number) checklist. 
- [x] If a new feature was added, a regression test for testing the new feature is added. 
 
### Testing

* How were these changes tested? Regression test on NOAA HPC
* Are the changes covered by regression tests?  Yes
* Have the matrix regression tests been run? Yes, Intel on NOAA HPC 

These two tests are not expected. Aron has them identical with Intel and Ali has them different on both Orion and Hera using Intel. @thesser1  could you test this test on ERDC HPC or your computer? One does not need mpi.

```
ww3_tr1/./work                     (1 files differ)
ww3_tr1/./work_std                     (1 files differ)
```
pre-known different cases:
```
**********************************************************************
********************* non-identical cases ****************************
**********************************************************************
mww3_test_03/./work_PR2_UQ_MPI_d2                     (6 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (6 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (8 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (12 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (6 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (8 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (8 files differ)
mww3_test_07/./work_PR3_UQ                     (4 files differ)
ww3_tp2.10/./work_MPI_OMPH                     (6 files differ)
ww3_tp2.16/./work_MPI_OMPH                     (4 files differ)
ww3_ufs1.3/./work_a                     (2 files differ)
```
Expected changes in mod_def files for unstructured cases:
```
ww3_tp2.17/./work_b                     (1 files differ)
ww3_tp2.17/./work_mb                     (1 files differ)
ww3_tp2.17/./work_mc                     (1 files differ)
ww3_tp2.17/./work_a                     (1 files differ)
ww3_tp2.17/./work_ma1                     (1 files differ)
ww3_tp2.17/./work_c                     (1 files differ)
ww3_tp2.17/./work_ma                     (1 files differ)
ww3_tp2.17/./work_mc1                     (1 files differ)
ww3_tp2.18/./work_TIDE_MPI                     (1 files differ)
ww3_tp2.6/./work_ST4                     (1 files differ)
ww3_tp2.6/./work_pdlib                     (1 files differ)
ww3_tp2.6/./work_ST0                     (1 files differ)
ww3_tp2.7/./work_ST0                     (1 files differ)
ww3_ts4/./work_ug_MPI                     (1 files differ)
```

* Please indicate the expected changes in the regression test output ([Note the known list of non-identical tests](https://github.com/NOAA-EMC/WW3/wiki/How-to-use-matrix.comp-to-compare-regtests-with-master#4-look-at-results)).
* Please provide the summary output of matrix.comp (_matrix.Diff.txt_, _matrixCompFull.txt_ and _matrixCompSummary.txt_):

[matrixCompFull.txt](https://github.com/NOAA-EMC/WW3/files/7687927/matrixCompFull.txt)
[matrixCompSummary.txt](https://github.com/NOAA-EMC/WW3/files/7687928/matrixCompSummary.txt)
[matrixDiff.txt](https://github.com/NOAA-EMC/WW3/files/7687929/matrixDiff.txt)

### CODE REVIEW

**Comment by JessicaMeixner-NOAA** [comment]
@aronroland Thanks for these nice updates - I can definitely see how this could impact any high-resolution case.  I'm curious if you or @aliabdolali looked into the impact of this change (beyond not changing answers) for structured multi-grid test cases that are already memory intensive, for example, the fully resolution test case example in ww3_ufs1.1.  Does this dramatically affect the memory usage for that test?  This is important information as double versus single precision is not an option in the way this change is introduced.

File: model/bin/comp.ite [hunk 1]
```diff
@@ -0,0 +1,213 @@
+#!/bin/bash
```
Comment (JessicaMeixner-NOAA) [inline model/bin/comp.ite]: All other comp/link platform specific files have been removed from the WW3 develop branch.  Should these be included in this PR or removed?

File: model/src/w3gdatmd.F90 [hunk 2]
```diff
@@ -2392,7 +2393,10 @@ SUBROUTINE W3SETG ( IMOD, NDSE, NDST )
       MAXX     => GRIDS(IMOD)%MAXX
       MAXY     => GRIDS(IMOD)%MAXY
       DXYMAX   => GRIDS(IMOD)%DXYMAX      
-
+      XGRD   => GRIDS(IMOD)%XGRD
+      YGRD   => GRIDS(IMOD)%YGRD
+      ZB     => GRIDS(IMOD)%ZB
+!AR: Todo: What ist he purpose of GINIT???
```
Comment (JessicaMeixner-NOAA) [inline model/src/w3gdatmd.F90]: This comment as is does not make sense.  Should this be removed or expanded on?

File: model/src/w3gridmd.F90 [hunk 3]
```diff
@@ -4949,14 +4949,7 @@ SUBROUTINE W3GRID()
...
       ELSE 
-!
-!FA:  This distinction  between structured and unstructured
-! should be removed when XYB is replaced by XGRD and YGRD
-!
-        DO IX=1, NX
-          XGRD(:,IX) = XYB(IX,1)
-          YGRD(:,IX) = XYB(IX,2)
-          END DO
+!AR: This is already done in readmsh
```
Comment (JessicaMeixner-NOAA) [inline model/src/w3gridmd.F90]: I'm a little confused by this comment.  Should it still be here?  If so, can you expand a little on this?

File: model/src/wmgridmd.F90 [hunk 4]
```diff
@@ -4166,7 +4172,10 @@ SUBROUTINE WMGEQL
                           MAPODI(IY  ,IX+1) .GE. 0. ) ) THEN
                 MASKI(IY,IX) = .TRUE.
 #ifdef W3_O12
-                IF ( IMPROC.EQ.NMPERR ) WRITE (MDSE,1020) I, IX, IY
+                IF ( IMPROC.EQ.NMPERR ) THEN
+                  WRITE (MDSE,1020) I, IX, IY
+                  WRITE (*,*) 'TESTING 4177 ERROR 1020', I, IX, IY
```
Comment (JessicaMeixner-NOAA) [inline model/src/wmgridmd.F90]: Remove this line as it seems to be a repeat of the formatted write statement above?

File: model/src/wmgridmd.F90 [hunk 5]
```diff
@@ -605,9 +606,11 @@ SUBROUTINE WMGLOW ( FLRBPI )
                          OUTPTS(I)%OUT5%RDBPI, STAT=ISTAT )
             CHECK_DEALLOC_STATUS ( ISTAT )
             CYCLE
-          ELSE IF ( IBI .NE. OUTPTS(I)%OUT5%NBI ) THEN
-             CALL EXTCDE ( 1020 )
-          ENDIF
+        ELSE IF ( IBI .NE. OUTPTS(I)%OUT5%NBI ) THEN
```
Comment (JessicaMeixner-NOAA) [inline model/src/wmgridmd.F90]: Revert to original code?

File: model/src/wmgridmd.F90 [hunk 6]
```diff
@@ -2966,8 +2969,10 @@ SUBROUTINE WMGHGH
                               NLOC_OM   = NLOC_OM + 1
 ! Notes: check here that we are sufficiently dimensioned.
                               IF ( NLOC_OM .GT. NLMAX ) THEN
-                                 IF ( IMPROC.EQ.NMPERR ) WRITE (MDSE,1020)
-                                 CALL EXTCDE(1020)
+                                 IF ( IMPROC.EQ.NMPERR ) WRITE (MDSE,*)
```
Comment (JessicaMeixner-NOAA) [inline model/src/wmgridmd.F90]: Revert to original formatted write statement and error call?

File: model/src/wmgridmd.F90 [hunk 7]
```diff
@@ -4166,7 +4172,10 @@ SUBROUTINE WMGEQL
                           MAPODI(IY  ,IX+1) .GE. 0. ) ) THEN
                 MASKI(IY,IX) = .TRUE.
 #ifdef W3_O12
-                IF ( IMPROC.EQ.NMPERR ) WRITE (MDSE,1020) I, IX, IY
+                IF ( IMPROC.EQ.NMPERR ) THEN
```
Comment (JessicaMeixner-NOAA) [inline model/src/wmgridmd.F90]: Revert to original write statement ?

File: model/src/ww3_ounf.F90 [hunk 8]
```diff
@@ -2494,6 +2494,7 @@ SUBROUTINE W3EXNC ( NX, NY, IX1, IXN, IY1, IYN, NSEA,             &
 
               ! If unstructured mesh
               IF (GTYPE.EQ.UNGTYPE) THEN
+!AR: for what do we need the array LON and LAT?!?!!?
```
Comment (JessicaMeixner-NOAA) [inline model/src/ww3_ounf.F90]: Clean up comment?

File: model/src/ww3_ounf.F90 [hunk 9]
```diff
@@ -3410,6 +3411,7 @@ SUBROUTINE W3EXNC ( NX, NY, IX1, IXN, IY1, IYN, NSEA,             &
       DEALLOCATE(MX1, MXX, MYY, MXY, MAPOUT)
       DEALLOCATE(MX1R, MXXR, MYYR, MXYR)
       DEALLOCATE(AUX1)
+!AR: This needs to be cleaned for the scalability part ..
```
Comment (JessicaMeixner-NOAA) [inline model/src/ww3_ounf.F90]: Clean up comment?

File: model/src/w3gridmd.F90 (hunk 3 above)
Comment (JessicaMeixner-NOAA) [inline model/src/w3gridmd.F90]: I'm a little confused by this comment.  Should it still be here?  If so, can you expand a little on this?

**Review by JessicaMeixner-NOAA — CHANGES_REQUESTED** [review]
Is there a reason that build_utils.sh and ft2src.sh were changed?  This seems unrelated to this PR along with the additional comp/link files.

**Comment by JessicaMeixner-NOAA** [comment]
@aliabdolali @aronroland there are still a few unresolved issues with this PR from the comments I made in the review last week.  There are also the open question as to why the permissions are being changed on a few files.  Lastly, there is the question on the impact of the memory since this is not added as an option. Please let me know if I need to look into that myself for impacts to the operations of the global systems here at NCEP.

**Comment by aliabdolali** [comment]
> @aliabdolali @aronroland there are still a few unresolved issues with this PR from the comments I made in the review last week. There are also the open question as to why the permissions are being changed on a few files. Lastly, there is the question on the impact of the memory since this is not added as an option. Please let me know if I need to look into that myself for impacts to the operations of the global systems here at NCEP.

@JessicaMeixner-NOAA the way @aronroland and I isolated and tested the scal_xyb does not affect the runtime and performance/outputs of the structured part of the code. As a sanity check, I checked the ufs1.* cases and there is no diff in the runtime. The reason we isolated the work is even one single change in the unstructured case made some changes in the structured part which sometimes takes hours to track and understand, sometimes impossible, therefore delaying the scalability reintegration and further development.
Aron can comment more here.

**Comment by JessicaMeixner-NOAA** [comment]
@aliabdolali it's great to hear runtime is unaffected.  How about the memory usage in the ufs1.1 with the high resolution grids? Again, if I need to do this work, I can I just need to know.

**Comment by aronroland** [comment]
@JessicaMeixner-NOAA <https://github.com/JessicaMeixner-NOAA> , @aliabdolali <https://github.com/aliabdolali> ,

 

Sorry for the late reply. I was busy this week and will have just time at the weekend to clean the code and reply to the comments from  <https://github.com/JessicaMeixner-NOAA> @JessicaMeixner-NOAA, many thanks for this in advance. Since we have deleted one array I do not expect big surprise. How do you  <https://github.com/JessicaMeixner-NOAA> @JessicaMeixner-NOAA check memory usage? 

 

Cheers

 

Aron

 

Von: Ali.Abdolali ***@***.***> 
Gesendet: Mittwoch, 15. Dezember 2021 15:53
An: NOAA-EMC/WW3 ***@***.***>
Cc: Aron Roland ***@***.***>; Mention ***@***.***>
Betreff: Re: [NOAA-EMC/WW3] Scal xyb (PR #556)

 

@aliabdolali <https://github.com/aliabdolali>  @aronroland <https://github.com/aronroland>  there are still a few unresolved issues with this PR from the comments I made in the review last week. There are also the open question as to why the permissions are being changed on a few files. Lastly, there is the question on the impact of the memory since this is not added as an option. Please let me know if I need to look into that myself for impacts to the operations of the global systems here at NCEP.

@JessicaMeixner-NOAA <https://github.com/JessicaMeixner-NOAA>  the way @aronroland <https://github.com/aronroland>  and I isolated and tested the scal_xyb does not affect the runtime and performance/outputs of the structured part of the code. As a sanity check, I checked the ufs1.* cases and there is no diff in the runtime. The reason we isolated the work is even one single change in the unstructured case made some changes in the structured part which sometimes takes hours to track and understand, sometimes impossible, therefore delaying the scalability reintegration and further development.
Aron can comment more here.

**Comment by JessicaMeixner-NOAA** [comment]
@aronroland there are a variety of ways to check the memory, including using the MEMCHECK switch you implemented.  Some HPCs give memory output and we can also check the resulting size of files.   Looks like I should just do these checks myself at this point.  I hope to have information for you in a week if not before.

**Comment by aronroland** [comment]
Hi Jessica, 

 

ok, so I guess that Ali and me can do that as well, as he is back. I would appreciate if could have a discussion with you and Ali in order to organize further steps. 

 

Cheers

 

Aron

 

Von: Jessica Meixner ***@***.***> 
Gesendet: Mittwoch, 15. Dezember 2021 16:17
An: NOAA-EMC/WW3 ***@***.***>
Cc: Aron Roland ***@***.***>; Mention ***@***.***>
Betreff: Re: [NOAA-EMC/WW3] Scal xyb (PR #556)

 

@aronroland <https://github.com/aronroland>  there are a variety of ways to check the memory, including using the MEMCHECK switch you implemented. Some HPCs give memory output and we can also check the resulting size of files. Looks like I should just do these checks myself at this point. I hope to have information for you in a week if not before.

**Comment by JessicaMeixner-NOAA** [comment]
For the ww3_tr1 files that are different, it's the mod_def files.  I wonder if something is written to the mod_def files that are changed by the changes made here or alternatively if this is a case of the memory issues that we seem to occasionally get.  One question would be if we run this branch twice, does that test give different answers?

**Comment by aronroland** [comment]
Dear Jessica, 

 

this is a bit strange stuff. So basically the mod_def file is different for Ali not for me and I cannot reproduce the difference, for which the final run results are the same. As we have looked at Ali’s binary difference using vdiffbin, the results have very tiny difference. 

 

I would have fixed it but I cannot reproduce it since using the intel compiler in our system (ifort version 2021.2.0), I could not find any difference in the binary. 

 

Code wise there should not be any difference but u never know. So what we wanted to do is to push this to Ty and see what he gets. I can also check on the French super computer. Hunting the differences down for the mod_def, would basically now result for me to introduce some possibility to write everything out in ascii and see where it differs. 

 

We will discuss this issue more on Monday I guess. 

 

Cheers

 

Aron

 

Von: Jessica Meixner ***@***.***> 
Gesendet: Donnerstag, 16. Dezember 2021 23:32
An: NOAA-EMC/WW3 ***@***.***>
Cc: Aron Roland ***@***.***>; Mention ***@***.***>
Betreff: Re: [NOAA-EMC/WW3] Scal xyb (PR #556)

 

For the ww3_tr1 files that are different, it's the mod_def files. I wonder if something is written to the mod_def files that are changed by the changes made here or alternatively if this is a case of the memory issues that we seem to occasionally get. One question would be if we run this branch twice, does that test give different answers?

**Comment by ukmo-ccbunney** [comment]
Hi @JessicaMeixner-NOAA and @aronroland 
Regarding he differences you are seeing in the mod_def.ww3 file for ww3_tr1, I have encountered this before and tracked it down to an uninitialized variable. See https://github.com/NOAA-EMC/WW3/issues/425

Seeing this jogged my memory - I thought a fix had been merged for that Issue, but I don't believe it has.
Perhaps you could test whether it fixes your problem (the details of the fix are in the issue) and I will raise a PR for it.

Cheers.

**Comment by aronroland** [comment]
This is great that u have reported that. Many thanks for this!

**Comment by JessicaMeixner-NOAA** [comment]
The size of the output files did not change, additionally the total memory usage on all of the ww3_ufs* tests increase with this PR, but not so much so that I think this warrants needing to put in an option for this variable.  If anyone is concerned about total memory usage for their applications please test this PR and make sure it's okay (@mickaelaccensi @ukmo-ccbunney & anyone else).

**Comment by aronroland** [comment]
Hi Jessica, 

 

can u please specify how u have tested the memory usage and post the certain numbers u have obtained? 

 

Cheers

 

Aron

 

Von: Jessica Meixner ***@***.***> 
Gesendet: Freitag, 17. Dezember 2021 17:15
An: NOAA-EMC/WW3 ***@***.***>
Cc: Aron Roland ***@***.***>; Mention ***@***.***>
Betreff: Re: [NOAA-EMC/WW3] Scal xyb (PR #556)

 

The size of the output files did not change, additionally the total memory usage on all of the ww3_ufs* tests increase with this PR, but not so much so that I think this warrants needing to put in an option for this variable. If anyone is concerned about total memory usage for their applications please test this PR and make sure it's okay ***@***.*** <https://github.com/mickaelaccensi>  @ukmo-ccbunney <https://github.com/ukmo-ccbunney>  & anyone else).

**Comment by JessicaMeixner-NOAA** [comment]
@aronroland I looked at the resulting file sizes and I used slurm commands to see the total memory usage of the regression test jobs for the ww3_ufs* tests.  The increase in total memory used was about .1GB or less. I didn't save output.

**Comment by aliabdolali** [comment]
> For the ww3_tr1 files that are different, it's the mod_def files. I wonder if something is written to the mod_def files that are changed by the changes made here or alternatively if this is a case of the memory issues that we seem to occasionally get. One question would be if we run this branch twice, does that test give different answers?

I checked it this morning, running tr1 twice ended with a different mod_def :(

File: model/bin/comp.ite (hunk 1 above)
Comment (aronroland) [reply model/bin/comp.ite]: Hi  @JessicaMeixner-NOAA this will be remove.

File: model/src/w3gridmd.F90 (hunk 3 above)
Comment (aronroland) [reply model/src/w3gridmd.F90]: This will be removed. Leftover from the merge.

File: model/src/wmgridmd.F90 (hunk 4 above)
Comment (aronroland) [reply model/src/wmgridmd.F90]: Removed!

File: model/src/wmgridmd.F90 (hunk 6 above)
Comment (aronroland) [reply model/src/wmgridmd.F90]: Reverted!

File: model/src/ww3_ounf.F90 (hunk 9 above)
Comment (aronroland) [reply model/src/ww3_ounf.F90]: Cleaned!

File: model/src/w3gridmd.F90 (hunk 3 above)
Comment (aronroland) [reply model/src/w3gridmd.F90]: Removed!

**Comment by aliabdolali** [comment]
After ww3_tr1 fix and resolving comments, I reran the regtests and here is the summary:

pre-known different cases:
```
**********************************************************************
********************* non-identical cases ****************************
**********************************************************************
mww3_test_03/./work_PR2_UQ_MPI_d2                     (8 files differ)
mww3_test_03/./work_PR2_UNO_MPI_d2                     (6 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2_c                     (8 files differ)
mww3_test_03/./work_PR1_MPI_d2                     (10 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2_c                     (8 files differ)
mww3_test_03/./work_PR3_UNO_MPI_d2                     (8 files differ)
mww3_test_03/./work_PR3_UQ_MPI_d2                     (8 files differ)
mww3_test_07/./work_PR3_UQ                     (1 files differ)
ww3_tp2.10/./work_MPI_OMPH                     (7 files differ)
ww3_tp2.16/./work_MPI_OMPH                     (5 files differ)
ww3_ufs1.3/./work_a                     (1 files differ)
```

Expected changes in mod_def files for unstructured cases:
```
ww3_tp2.17/./work_b                     (1 files differ)
ww3_tp2.17/./work_mb                     (1 files differ)
ww3_tp2.17/./work_mc                     (1 files differ)
ww3_tp2.17/./work_a                     (1 files differ)
ww3_tp2.17/./work_ma1                     (1 files differ)
ww3_tp2.17/./work_c                     (1 files differ)
ww3_tp2.17/./work_ma                     (1 files differ)
ww3_tp2.17/./work_mc1                     (1 files differ)
ww3_tp2.21/./work_b                     (1 files differ)
ww3_tp2.21/./work_mb                     (1 files differ)
ww3_tp2.21/./work_a                     (1 files differ)
ww3_tp2.21/./work_ma                     (1 files differ)
ww3_tp2.6/./work_ST4                     (1 files differ)
ww3_tp2.6/./work_pdlib                     (1 files differ)
ww3_tp2.6/./work_ST0                     (1 files differ)
ww3_tp2.7/./work_ST0                     (1 files differ)
ww3_ts4/./work_ug_MPI                     (1 files differ)
```

**Review by JessicaMeixner-NOAA — APPROVED** [review]

**Pull request merged**